### ISSUE TITLE
GLPI 9.4.0 - Automatic Action closeticket stuck in running state.

### ISSUE THREAD

**Issue by peraun** [glpi-project/glpi#5558]
**Symptoms:**

The Automatic Action closeticket is stuck in running status even if I clean the last run and force it to execute. Also when I click Execute to force it will redirect me to a random ticket page.
I'm also receiving an email reporting:

The following automatic actions are in error. They require intervention.

closeticket: Automatic Tickets Closing

1. Go to 'Setup > Automatic Actions > select closeticket'
2. In the "Last run" and "Next Run" If I click on the X to clean the last run and next it would be expected to wait and run the next time scheduled.
3. If I wait it will run indeed but the field "Next Run" keeps displaying Running and even when I force it to run it will keep the same behavior.

PS: The Run Mode of closeticket is set to CLI.

**Expected behavior:**
It would run and close automatically the solved tickets as expected and display the date it would run again.

**Logs:**
If I clean the status and force it to run at "Setup > Automatic Actions > select closeticket" cron.log will show:

2019-03-08 14:38:18 [371@web01]
Internal #1: Launch closeticket

If I run it via cli by running "/usr/bin/php /usr/share/glpi/front/cron.php" it will show:

2019-03-08 14:41:49 [@web01]
External #1: Launch closeticket

I couldn't find any other error or mention to the cron in the GLPI logs or apache.

**Screenshots:**

In the attached Screenshot #1 you can see that when I force it to run it displays Running:
![2](https://user-images.githubusercontent.com/7115215/54047093-e09fd280-41b4-11e9-9988-5cef518f5268.png)

But when I go to Historical of the task it displays that the status was changed from Running to scheduled:
![1](https://user-images.githubusercontent.com/7115215/54047074-d251b680-41b4-11e9-85a8-e28dc2dc2a2f.png)

I'm running the GLPI version 9.4.0 on Ubuntu 18.04 both in the web server and mysql server. Both are up to date too.

**Additional context:**

I've recorded a little video showing the problem and I'll describe the timeline of the video:

https://www.dropbox.com/s/mw2cx7pisi194kq/aut-action-close-mp4.m4v?dl=0

1sec - 5sec: It shows me cleaning the status and clicking in execute for force it to run;

6sec - 11sec: It shows that as soon as I click Execute it redirects the page to a random ticket and in the Satisfaction tab;

12sec - 38sec: I click go back to the closeticket page and reload it to show that the running status is still stuck. Then I change the "Run Frequency" to 1 minute, clean the status again and reload the page to wait it to run automatically;

46sec - : I was monitoring the cron.log and as soon as the log displays the action runs I reload the page and it will keep the status Running again.

**Comment by arhipovdv** [comment]
Have absolutely same behavior with mailgate automatic action. no logs, no errors, stuck in running. Manual reciver start (setup - receivers -actions - Get email tickets now) work, atomatic cli - not...

**Comment by cedric-anne** [comment]
@peraun ,

I tried to reproduce but cannot.
1. Are you running the cron using the same system user as the one used to serve pages (www-data for example) ?
2. Are you using multiple entiteis ? If yes, could you give the configuration of your entities that has tickets to close ?

@arhipovdv ,
You are talking about a different cron/use case. Please create another ticket with all requested informations if you want your issue treated correclty.

Regards

**Comment by peraun** [comment]
@cedric-anne thanks for the reply! Here the answers:
1-  I guess is that it's using the same user to serve pages (www-data). As the screenshot shows I'm running in CLI mode but as GLPI doesn't work as well.
(https://user-images.githubusercontent.com/7115215/54047093-e09fd280-41b4-11e9-9988-5cef518f5268.png)
2- I'm using multiple entities but they are child of the root. Is the configuration of tickets to close you are asking the **Automatic closing of solved tickets after**  option? If it's the configurationfor both child entities I have is **Inheritance of the Parent Entity**
![Screen Shot 2019-03-11 at 11 43 46](https://user-images.githubusercontent.com/7115215/54132633-55f8e680-43f3-11e9-8d11-355d79c259a6.png)

I forgot to mention that if I click Execute (it does what the video shows), I get an email saying that a random ticket was closed but the closing date doesn't match the specified date to automatic close.
![Screen Shot 2019-03-11 at 11 50 06](https://user-images.githubusercontent.com/7115215/54133131-49c15900-43f4-11e9-80a0-9e69bed02eeb.png)

**Comment by cedric-anne** [comment]
@peraun 

Could you apply following patch and run the cron from cli to see if anything is displayed ?

```diff
diff --git a/front/cron.php b/front/cron.php
index 0f29577376..5e01a48bc9 100644
--- a/front/cron.php
+++ b/front/cron.php
@@ -44,6 +44,11 @@ if (!is_writable(GLPI_LOCK_DIR)) {
    exit (1);
 }
 
+if (isCommandLine()) {
+   // Force debug mode in command line mode to output errors
+   Toolbox::setDebugMode(Session::DEBUG_MODE);
+}
+
 if (!isCommandLine()) {
    //The advantage of using background-image is that cron is called in a separate
    //request and thus does not slow down output of the main page as it would if called
```

Regards

**Comment by peraun** [comment]
Sure! Here what logged:
First it didn't executed any action because the task was still running, then after I stopped it and cleaned the last run and run it from cli it displayed:
![Screen Shot 2019-03-12 at 07 55 07](https://user-images.githubusercontent.com/7115215/54194998-58fbe180-449c-11e9-8f49-087457a20aef.png)
Where is "Um e-mail para xxx foi adicionado para a fila" you read "An email to xxx were added to the queue"

**Comment by cedric-anne** [comment]
Is the task still marked as "running" ?
If yes, as no error is displayed, I really do not know what can explain this.

**Comment by arhipovdv** [comment]
> @peraun
> 
> Could you apply following patch and run the cron from cli to see if anything is displayed ?
> 
> ```diff
> diff --git a/front/cron.php b/front/cron.php
> index 0f29577376..5e01a48bc9 100644
> --- a/front/cron.php
> +++ b/front/cron.php
> @@ -44,6 +44,11 @@ if (!is_writable(GLPI_LOCK_DIR)) {
>     exit (1);
>  }
>  
> +if (isCommandLine()) {
> +   // Force debug mode in command line mode to output errors
> +   Toolbox::setDebugMode(Session::DEBUG_MODE);
> +}
> +
>  if (!isCommandLine()) {
>     //The advantage of using background-image is that cron is called in a separate
>     //request and thus does not slow down output of the main page as it would if called
> ```
> 
> Regards

its help me too! in debug message, when mailgate runs, it was an error with https://github.com/pluginsGLPI/tag plugin. Plugin was installed, but disabled. sorry cant post it here (cleaned the output) but it was something like: cant find tag filed in ticket card. When i uninstall it and remove it folder - mailgate runs normal!
Thnx!

**Comment by peraun** [comment]
Only closedticket is marked as running, every time I run the task it persist as running. Only after I click in the x it stops but as soon as the next time cron runs it keep the running status again.

**Comment by cedric-anne** [comment]
@peraun ,

1. Have you some active plugins ?
2. How many entities do you have ? And how many tickets should be closed by this cron ?

Regards

**Comment by peraun** [comment]
@cedric-anne ,
1. Here the plugins list:
![Screen Shot 2019-03-12 at 10 31 11](https://user-images.githubusercontent.com/7115215/54203949-01687080-44b2-11e9-89e5-1ff0b4087bfa.png)
2. a- I have a root entity and 2 child entities. 
    b- Right now I have 188 tickets marked as solved before 7 days from today. Tickets which should have the status closed as the Automatic Closing Configuration is configured to 7 days for the root entity and child entities.
![Screen Shot 2019-03-12 at 10 40 14](https://user-images.githubusercontent.com/7115215/54204573-3d500580-44b3-11e9-9ee9-7c360daaebb2.png)

**Comment by cedric-anne** [comment]
The cron parses entities separately, so if your ticket `0020689` is in another entity than others, it is normal that its processing is not done in this sequence.

Is it possible for you to test using the current nightly build and without any active plugin ?

Regards

**Comment by kosmic69** [comment]
Hi,

We have the same problem and I already posted my symptoms on the glpi forum https://forum.glpi-project.org/viewtopic.php?id=168029. Another people has the same problem https://forum.glpi-project.org/viewtopic.php?id=167967.

The problem is not the cron job but only the close ticket action who hang.
The cron job is blocked because of the close ticket action.

Thanks for the bug repair.

**Comment by supports2l** [comment]
Hi,

I had the same problem with my IIS server.
Additionaly, when I clicked on "Execute" for this automatic action, I was redirected to a white page with only this html 

&lt;head&gt;&lt;title&gt;Document Moved&lt;/title&gt;&lt;/head&gt;
&lt;body&gt;&lt;h1&gt;Object Moved&lt;/h1&gt;This document may be found &lt;a HREF=""&gt;here&lt;/a&gt;

I wrote this URL Rewrite rule in IIS, 
```
   <outboundRules>
	<rule name="HTTP201" preCondition="isHTTP201-1" enabled="true" stopProcessing="false">
		 <match filterByTags="None" pattern="^&lt;head>(.|\r\n|\n)*?&lt;\/body>(\{.*)$" />
		 <action type="Rewrite" value="{R:2}" />
	</rule>
	<preConditions>
		<preCondition name="isHTTP201-1">
		   <add input="{RESPONSE_CONTENT_TYPE}" pattern="^text/html" />
		   <add input="{RESPONSE_STATUS}" pattern="302" />
		   <add input="{RESPONSE_LOCATION}" pattern="^https?://.*?/crontask.form.php/" />
		</preCondition>
	 </preConditions>
   </outboundRules>
```

Now, I can start the task manually and the job is done.

But the job is still stuck (running) when called by cron (run mode CLI).

**Comment by peraun** [comment]
> The cron parses entities separately, so if your ticket `0020689` is in another entity than others, it is normal that its processing is not done in this sequence.
> 
> Is it possible for you to test using the current [nightly build](https://forge.glpi-project.org/tarballs/9.4-bugfixes.tar.gz) and without any active plugin ?
> 
> Regards

@cedric-anne Sorry for my delay replying but only today I could try.
I did a clean install of glpi 9.4.1 (9.4.1.1) and restore the db from my production glpi and the automatic action worked, running it both via web and cli worked and the action wasn't stuck.
I'll try updating my production to see if it fix the problem.

Best Regards,
Paulo

**Comment by kosmic69** [comment]
> Hi,
> 
> We have the same problem and I already posted my symptoms on the glpi forum https://forum.glpi-project.org/viewtopic.php?id=168029. Another people has the same problem https://forum.glpi-project.org/viewtopic.php?id=167967.
> 
> The problem is not the cron job but only the close ticket action who hang.
> The cron job is blocked because of the close ticket action.
> 
> Thanks for the bug repair.

With the version 9.4.1.1 it works for us also.

**Comment by cedric-anne** [comment]
Hi,

Thanks for the feedback. I close the issue as it seems to be solved.

Regards

**Comment by peraun** [comment]
I just updated my production glpi and it fixed the problem. Thank you all!

**Comment by GeneralVini** [comment]
same error on glpi 9.4.5

**Comment by simplifyhealthcare** [comment]
same error with GLPI 9.5.2 

crontask.form.php?id=9  page hangs overall application.

**Comment by lolokai** [comment]
I have the same error
